File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/audio/AudioTypeChain.java
package com.mapbox.services.android.navigation.v5.navigation.metrics.audio;

public class AudioTypeChain {

  public AudioTypeResolver setup() {
    AudioTypeResolver unknown = new UnknownAudioType();
    AudioTypeResolver speaker = new SpeakerAudioType();
    speaker.nextChain(unknown);
    AudioTypeResolver headphones = new HeadphonesAudioType();
    headphones.nextChain(speaker);
    AudioTypeResolver rootOfTheChain = new BluetoothAudioType();
    rootOfTheChain.nextChain(headphones);

    return rootOfTheChain;
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/metrics/audio/AudioTypeChain.kt
package com.mapbox.services.android.navigation.v5.internal.navigation.metrics.audio

internal class AudioTypeChain {

    fun setup(
        unknownAudioType: AudioTypeResolver.Unknown = AudioTypeResolver.Unknown(),
        speakerAudioType: AudioTypeResolver.Speaker = AudioTypeResolver.Speaker(),
        headphonesAudioType: AudioTypeResolver.Headphones = AudioTypeResolver.Headphones(),
        bluetoothAudioType: AudioTypeResolver.Bluetooth = AudioTypeResolver.Bluetooth()
    ): AudioTypeResolver {
        speakerAudioType.nextChain(unknownAudioType)
        headphonesAudioType.nextChain(speakerAudioType)
        bluetoothAudioType.nextChain(headphonesAudioType)
        return bluetoothAudioType
    }
}
